massa-models/src/address.rs

```diff
@@ -179,202 +276,606 @@ impl FromStr for Address {
+    pub fn get_thread(&self, thread_count: u8) -> Option<u8> {
+        match self {
+            Address::User(addr) => Some(addr.get_thread(thread_count)),
+            Address::SC(_addr) => None,
+        }
+    }
+
+    /// Computes the address associated with the given public key.
+    /// Depends on the Public Key version
+    pub fn from_public_key(public_key: &PublicKey) -> Self {
+        Address::User(UserAddress::from_public_key(public_key))
+    }
+
+    /// Serialize the address as bytes. Includes the type and version prefixes
+    pub fn to_prefixed_bytes(self) -> Vec<u8> {
+        match self {
+            Address::User(addr) => addr.to_prefixed_bytes(),
+            Address::SC(addr) => addr.to_prefixed_bytes(),
+        }
+    }
+}
+
+impl UserAddress {
+    /// Gets the associated thread. Depends on the `thread_count`
+    fn get_thread(&self, thread_count: u8) -> u8 {
+        match self {
+            UserAddress::UserAddressV0(addr) => addr.get_thread(thread_count),
+            UserAddress::UserAddressV1(addr) => addr.get_thread(thread_count),
+        }
+    }
+
+    /// Computes the address associated with the given public key.
+    /// Depends on the Public Key version
+    fn from_public_key(public_key: &PublicKey) -> Self {
+        //BuildVersions!(version, from_public_key(public_key), ModelsError::InvalidVersionError(format!("Address version {} doesn't exist.", version)));
+
+        match public_key {
+            PublicKey::PublicKeyV0(pk) => {
+                UserAddressVariant!["0"](<UserAddress!["0"]>::from_public_key(pk))
+            }
+            PublicKey::PublicKeyV1(pk) => {
+                UserAddressVariant!["1"](<UserAddress!["1"]>::from_public_key(pk))
+            }
+        }
+    }
+
+    fn from_str_without_prefixed_type(s: &str) -> Result<Self, ModelsError> {
+        let decoded_bs58_check = bs58::decode(s)
             .with_check(None)
             .into_vec()
             .map_err(|_| ModelsError::AddressParseError)?;
-
-        // extract the version
         let u64_deserializer = U64VarIntDeserializer::new(Included(0), Included(u64::MAX));
-        let (rest, _version) = u64_deserializer
+        let (rest, version) = u64_deserializer
             .deserialize::<DeserializeError>(&decoded_bs58_check[..])
             .map_err(|_| ModelsError::AddressParseError)?;
 
-        // ...and package it up
-        let res = UserAddress(Hash::from_bytes(
-            rest.try_into()
-                .map_err(|_| ModelsError::AddressParseError)?,
-        ));
+        match version {
+            <UserAddress!["0"]>::VERSION => Ok(UserAddressVariant!["0"](
+                <UserAddress!["0"]>::from_bytes_without_version(rest)?,
+            )),
+            <UserAddress!["1"]>::VERSION => Ok(UserAddressVariant!["1"](
+                <UserAddress!["1"]>::from_bytes_without_version(rest)?,
+            )),
+            _ => Err(ModelsError::AddressParseError),
+        }
+    }
```

Comment:
good candidate to use the `nom::alt` parser. We can pair together to make it work